pkg/dns/serviceresolver.go

```diff
@@ -196,6 +201,16 @@ func (b *ServiceResolver) Records(dnsName string, exact bool) ([]msg.Service, er
 			return nil, errNoSuchName
 		}
 
+		hostnameMappings := noHostnameMappings
+		if savedHostnames := endpoints.Annotations[kendpoints.PodHostnamesAnnotation]; len(savedHostnames) > 0 {
+			mapped := make(map[string]kendpoints.HostRecord)
+			if err = json.Unmarshal([]byte(savedHostnames), &mapped); err == nil {
+				hostnameMappings = mapped
+			}
+		}
+
+		matchHostname := len(segments) > 3 && !hasAllPrefixedSegments(segments[3:4], "_")
```

Comment:
The godoc at the top of this function should be updated to clarify that we now either return a hash of the IP or a hostname value.  Additionally, it would be very useful to people looking at/reviewing this code in the future if there were comments added before each "branch point"/conditional indicating which subquery they were matching (for instance, AFAICT, this one might read "matchHostname indicates we are matching <hostname>.<svc>.<ns>.endpoints, and not _endpoints.<svc>.<ns>.svc")